src/python/zquantum/qaoa/recursive_qaoa.py

```diff
@@ -202,17 +178,44 @@ def __call__(
         ):
             # If we didn't reach threshold `n_c`, we repeat the the above with the reduced
             # cost hamiltonian.
-            return self.__call__(reduced_cost_hamiltonian, new_qubit_map)
+
+            self._cost_hamiltonian = reduced_cost_hamiltonian
+            self._qubit_map = new_qubit_map
+
+            return self.minimize(cost_function_factory, initial_params, keep_history)
```

Comment:
I think that instead of doing recursion using `minimize` (which has a very specific signature), it would be better if we moved all that logic in a separate function that will be used for the recursion calls.
This will also allow to get rid of the whole section "variables are for keeping track of data across multiple recursions" in init, because we'll be able to pass them through calling that recursive function. Once this is done, we can think how to make it more elegant and compact, maybe through using some auxiliary data structures.

This will probably mean that `minimize` will become very thin, but that's good.

Reply:
Sounds good